src/net_processing.cpp

```diff
@@ -2858,36 +2872,40 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             return;
         }
 
-        LOCK(cs_main);
+        CInv inv;
 
-        const CBlockIndex* pindex = LookupBlockIndex(req.blockhash);
-        if (!pindex || !(pindex->nStatus & BLOCK_HAVE_DATA)) {
-            LogPrint(BCLog::NET, "Peer %d sent us a getblocktxn for a block we don't have\n", pfrom.GetId());
-            return;
-        }
+        {
+            LOCK(cs_main);
+
+            const CBlockIndex* pindex = LookupBlockIndex(req.blockhash);
+            if (!pindex || !(pindex->nStatus & BLOCK_HAVE_DATA)) {
+                LogPrint(BCLog::NET, "Peer %d sent us a getblocktxn for a block we don't have\n", pfrom.GetId());
+                return;
+            }
+
+            if (pindex->nHeight >= ::ChainActive().Height() - MAX_BLOCKTXN_DEPTH) {
+                CBlock block;
+                bool ret = ReadBlockFromDisk(block, pindex, m_chainparams.GetConsensus());
+                assert(ret);
+
+                SendBlockTransactions(pfrom, block, req);
+                return;
+            }
 
-        if (pindex->nHeight < ::ChainActive().Height() - MAX_BLOCKTXN_DEPTH) {
             // If an older block is requested (should never happen in practice,
```

Comment:
Rather than splitting this logic up (hoisting the `CInv` declaration above the code block, fetching the `type` inside the `cs_main` scope, and then putting the rest of the logic outside the cs_main scope), I recommend you keep all the logic together:

```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2872,8 +2872,6 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             return;
         }
 
-        CInv inv;
-
         {
             LOCK(cs_main);
 
@@ -2892,17 +2890,18 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
                 return;
             }
 
-            // If an older block is requested (should never happen in practice,
-            // but can happen in tests) send a block response instead of a
-            // blocktxn response. Sending a full block response instead of a
-            // small blocktxn response is preferable in the case where a peer
-            // might maliciously send lots of getblocktxn requests to trigger
-            // expensive disk reads, because it will require the peer to
-            // actually receive all the data read from disk over the network.
-            inv.type = State(pfrom.GetId())->fWantsCmpctWitness ? MSG_WITNESS_BLOCK : MSG_BLOCK;
         }
 
+        // If an older block is requested (should never happen in practice,
+        // but can happen in tests) send a block response instead of a
+        // blocktxn response. Sending a full block response instead of a
+        // small blocktxn response is preferable in the case where a peer
+        // might maliciously send lots of getblocktxn requests to trigger
+        // expensive disk reads, because it will require the peer to
+        // actually receive all the data read from disk over the network.
         LogPrint(BCLog::NET, "Peer %d sent us a getblocktxn for a block > %i deep\n", pfrom.GetId(), MAX_BLOCKTXN_DEPTH);
+        CInv inv;
+        inv.type = WITH_LOCK(cs_main, return State(pfrom.GetId())->fWantsCmpctWitness ? MSG_WITNESS_BLOCK : MSG_BLOCK);
         inv.hash = req.blockhash;
```

It's slightly wasteful that cs_main is released and then taken again, but I think it's ok because:

1. As the comment says, we don't expect this to happen except in tests
2. `fWantsCmpctWitness` should move to the `Peer` struct eventually, removing the need to take cs_main

Reply:
Agreed, this is nicer! Changed. 

Interestingly, your exact suggestion didn't work: I had to move the assignment to `inv.type` inside the `WITH_LOCK` macro. I got the following error the other way:

```
./sync.h:257:45: note: expanded from macro 'WITH_LOCK'
#define WITH_LOCK(cs, code) [&] { LOCK(cs); code; }()
                                            ^~~~
net_processing.cpp:2934:20: error: assigning to 'uint32_t' (aka 'unsigned int') from incompatible type 'void'
        inv.type = WITH_LOCK(cs_main, State(pfrom.GetId())->fWantsCmpctWitness ? MSG_WITNESS_BLOCK : MSG_BLOCK);
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sync.h:257:29: note: expanded from macro 'WITH_LOCK'
#define WITH_LOCK(cs, code) [&] { LOCK(cs); code; }()
                            ^~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
It looks like you missed out the `return` from inside the `WITH_LOCK` macro's second argument:

`        inv.type = WITH_LOCK(cs_main, return State(pfrom.GetId())->fWantsCmpctWitness ? MSG_WITNESS_BLOCK : MSG_BLOCK);`

But your way is also fine!